packages/i18n-calypso/src/number-formatters/number-format-currency/index.ts

```
@@ -0,0 +1,198 @@
+	}
+
+	return scaleNumberForPrecision( number, currencyPrecision );
+}
+
+/**
+ * Formats money with a given currency code.
+ *
+ * The currency will define the properties to use for this formatting, but
+ * those properties can be overridden using the options. Be careful when doing
+ * this.
+ *
+ * For currencies that include decimals, this will always return the amount
+ * with decimals included, even if those decimals are zeros. To exclude the
+ * zeros, use the `stripZeros` option. For example, the function will normally
+ * format `10.00` in `USD` as `$10.00` but when this option is true, it will
+ * return `$10` instead.
+ *
+ * Since rounding errors are common in floating point math, sometimes a price
+ * is provided as an integer in the smallest unit of a currency (eg: cents in
+ * USD or yen in JPY). Set the `isSmallestUnit` to change the function to
+ * operate on integer numbers instead. If this option is not set or false, the
+ * function will format the amount `1025` in `USD` as `$1,025.00`, but when the
+ * option is true, it will return `$10.25` instead.
+ *
+ * If the number is NaN, it will be treated as 0.
+ *
+ * If the currency code is not known, this will assume a default currency
+ * similar to USD.
+ *
+ * If `isSmallestUnit` is set and the number is not an integer, it will be
+ * rounded to an integer.
+ * @returns    {string}                  A formatted string.
+ */
+const numberFormatCurrency: NumberFormatCurrency = ( {
+	number,
+	browserSafeLocale,
+	currency,
+	stripZeros,
+	isSmallestUnit,
+	signForPositive,
+	geoLocation,
+	forceLatin,
+} ) => {
+	const validCurrency = getValidCurrency( currency, geoLocation );
+	const currencyOverride = getCurrencyOverride( validCurrency, geoLocation );
+	const currencyPrecision = getPrecisionForLocaleAndCurrency( browserSafeLocale, validCurrency );
+	const numberAsFloat = prepareNumberForFormatting(
+		number,
+		currencyPrecision ?? 0,
```

This can never be undefined because you are setting it as 3 in `getPrecisionForLocaleAndCurrency()`. Do we want to default to 3 or 0? (Sorry I should have noticed this in https://github.com/Automattic/wp-calypso/pull/99981/). It probably doesn't matter since this should rarely happen but I think that 0 might be slightly safer (?) since this is here for converting smallest units to floats. Actually, it's pretty risky no matter what (the difference is between displaying $1, $100, or $1000). Maybe we should error out if we can't tell the precision for the currency?

---

Ah I think this lingered from before updating the return value. It does sound "safer" to go with less than more, but honestly not sure.

Do you mean to `console.warn` or error out in full, like throw an actual error?

---

Hm... I'm not sure. 

Let's take an example. What would be the best user experience if some code said to format the smallest unit number `100` but we can't tell how many decimals are in that currency? 

If we choose 0, we would render it as something like `$100`. If we choose 2, we would render it as `$1.00`. If we choose 3, we would render it as `$0.10`.

If we display a console warning or error, the user won't see it and will see whichever of the above options we choose. Maybe we could display something like "Unknown" in this case?

If we throw an error, the page rendering may stop and (hopefully, if there's a fallback) display a message to contact support.

I think the best option in this situation is to fail loudly enough that support staff will be notified, so probably one of the second two options.

---

I agree, overall.

> If we display a console warning or error, the user won't see it and will see whichever of the above options we choose. Maybe we could display something like "Unknown" in this case?

I would go with a `null` return in such a case, and the consumer can deal with it. I think of `null` as the more standard "intentionally empty" state. However, that will change the signature, and it's likely that consumers of this API will have to deal with a potential null return in all cases (irrespective of this being an incredibly rare scenario). So, not sure I'd modify the signature to be of `string | null` return, or introduce an arbitrary variable like "unknown".

> I think the best option in this situation is to fail loudly enough that support staff will be notified, so probably either displaying "Unknown" (or something similar) or throwing a fatal.

This might be it. I am reading up on when these defaults may _not_ be set ([here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#significantdigitsfractiondigits_default_values)) and it sounds likely to be the case (unless we intentionally orchestrate things that way). It's also only relevant for `isSmallestUnit`, as otherwise forcing any default sounds ok 🤔

I've made an update throwing an error. Let me know what you think! I can incorporate the change to the base PR then.

---

> I've made an update throwing an error. Let me know what you think! I can incorporate the change to the base PR then.

It includes `isSmallestUnit` in the condition. I think it doesn't really matter, so happy to remove that and keep it simple. Let me know! :-)

---

Looks good to me! We can always revisit later but hopefully this is a situation that will never occur.

---

Sounds good. 

I'll ping you once PR is ready to give it a go with a few actual uses. Then I can incorporate things into https://github.com/Automattic/wp-calypso/pull/99981